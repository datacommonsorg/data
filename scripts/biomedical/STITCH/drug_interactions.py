# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
'''
Author: Mariam Benazouz
Date: 08/12/2022
Name: drug_interactions.py
Description: creates PubChem DCIDs for drug-drug interactions
@PATH_DRUG_INTERACTIONS   file path for drug-drug interactions file 
'''
import numpy as np
import pandas as pd
import sys

def pubchem_dcid(PATH): 
    """
    Create PubChem ID column for each chemical and their interaction as
    well as a PubChem interaction DCID 
    
    Parameters:
    ----------
    PATH: str
       path for the file containing data on drug-drug interactions
    Returns:
    --------
    Pandas dataframe 
    
    TEST:
    df_drug_interactions = pubchem_dcid(PATH_DRUG_INTERACTIONS)
    assert(type(df_chemicals_reformatted) == pd.core.frame.dataframe)
    """
    df_drug_interactions = pd.read_csv(PATH, sep = '\t')
    df_drug_interactions['chemical1']=df_drug_interactions[
        'chemical1'].replace(r'ID[a-z]?0*','ID',regex=True)
    df_drug_interactions['chemical2']=df_drug_interactions[
        'chemical2'].replace(r'ID[a-z]?0*','ID',regex=True)
    s1_le = df_drug_interactions['chemical1'] <= df_drug_interactions[
        'chemical2']
    df_drug_interactions['pubchem_combined'] = pd.concat((
        df_drug_interactions.loc[s1_le,'chemical1']+'_'+\
        df_drug_interactions.loc[s1_le,'chemical2'],
        df_drug_interactions.loc[~s1_le,'chemical2']+\
        '_'+df_drug_interactions.loc[~s1_le,'chemical1']))
    df_drug_interactions['pubchem_dcid']='chem/'+\
        df_drug_interactions['pubchem_combined']
    df_drug_interactions['chemical1']='chem/'+ \
        df_drug_interactions['chemical1']
    df_drug_interactions['chemical2']='chem/'+ \
        df_drug_interactions['chemical2']
    df_drug_interactions['sort'] = df_drug_interactions[
        'pubchem_combined'].str.extract('(\d+)', 
        expand=False).astype(int)
    df_drug_interactions.sort_values('sort', inplace = True, 
        ascending = True)
    df_drug_interactions.drop('sort', axis=1, inplace = True)
    df_drug_interactions.fillna('', inplace=True)
    return(df_drug_interactions)

def main():
    PATH_DRUG_INTERACTIONS = sys.argv[1]
    df_drug_interactions = pubchem_dcid(PATH_DRUG_INTERACTIONS)
    df_drug_interactions.to_csv('drug_interactions.csv', index=False)  

if __name__ == '__main__':
    main()