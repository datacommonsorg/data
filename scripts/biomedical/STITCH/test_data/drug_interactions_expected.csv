chemical1,chemical2,similarity,experimental,database,textmining,combined_score,pubchem_combined,pubchem_dcid
CID24759,CID24759,0,0,900,0,900,CID24759_CID24759,chem/CID24759_CID24759
CID91758695,CID107694,0,0,0,230,230,CID91758695_CID107694,chem/CID91758695_CID107694
CID91758695,CID11013287,0,0,0,230,230,CID91758695_CID11013287,chem/CID91758695_CID11013287
CID91758695,CID11980957,0,0,0,328,328,CID91758695_CID11980957,chem/CID91758695_CID11980957
CID91758695,CID13078,0,0,0,162,162,CID91758695_CID13078,chem/CID91758695_CID13078
CID91758695,CID13109,0,0,0,468,468,CID91758695_CID13109,chem/CID91758695_CID13109
CID91758695,CID167486,0,0,0,194,194,CID91758695_CID167486,chem/CID91758695_CID167486
CID91758695,CID24706,0,0,0,452,452,CID91758695_CID24706,chem/CID91758695_CID24706
CID91758695,CID24883487,0,0,0,271,271,CID91758695_CID24883487,chem/CID91758695_CID24883487
CID91758695,CID40973,0,0,0,505,505,CID91758695_CID40973,chem/CID91758695_CID40973
